File: sp/go/pingpong/system.go
package tdpnet

type System struct {
	*StandardSession
	Config SystemConfig
}

func NewSystem(config SystemConfig, connectNow bool) *System {
	s := &System{
		StandardSession: NewStandardSession(),
		Config:          config,
	}

	if connectNow {
		// In a real Go app, you would pass config.Security.CA to a tls.Config
		s.Connect(config.Host)
	}
	return s
}

func (s *System) OnConnected(timestamp int64) {
	// Automatically sign in upon connection
	s.Signin(s.Config.Credential)
}

// Placeholder for granted/denied logic
func (s *System) OnGranted(data interface{}) {}
func (s *System) OnDenied(data interface{})  {}

File: sp/rust/pingpong/system.rs
use crate::tdpnet::TDPnet; // From previous translation
use crate::config::Config;
use serde_json::json;

pub struct System {
    pub tdp: TDPnet,
    pub config: &'static Config,
}

impl System {
    pub async fn new(config: &'static Config, connect: bool) -> Self {
        let sys = Self {
            tdp: TDPnet::new(),
            config,
        };
        if connect {
            // Pass host and CA cert to the connection method
            let _ = sys.tdp.connect(config.host).await;
        }
        sys
    }

    pub fn signin(&self) {
        let cred = json!({
            "accesskey": self.config.credential.accesskey,
            "password": self.config.credential.password,
            "address": self.config.credential.address,
        });
        // Logic to send signin packet via tdp
    }
}